### PR TITLE
Fix notebook query cleaning

### LINKED ISSUE #11334: Custom Question builder, changing first step removes subsequent steps completely
**Describe the bug**
When using the Custom question builder (notebook mode), if there are a series of joins and processing steps, changing the first step completely removes the subsequent steps.

This essentially means you've got to get the first step 100% correct before moving forward. Part of the benefit of these notebook style editors is that you can try stuff then tweak it to correct issues, this reset is blocking that.

**Logs**
No output in any logs

**To Reproduce**
1. Go to Custom question screen
2. Setup as shown below (Orders > filter for Product category "Doohickey" > Sum Total by user ID > Join People > Count by People ID
image

3. In the first "Filter" section add category "Gizmo" and click the "Update Filter" button
4. Note the 2nd step of the question is removed (see screenshot) meaning I need to repeat those step from memory
image

**Expected behavior**
Subsequent processing steps should be retained when editing earlier steps. I understand that there could be dependencies between these steps but often that's not the case.

I expected to be able to add a new filter and not lose most of my work. Notebooks should enable exploratory analysis, and allow you to build up a complex series of operations. Resetting these steps prevents that.

**Screenshots**

**Information about your Metabase Installation:**

- Your browser and the version: Chrome (v77.0.3865.120)
- Your operating system: Windows 7
- Your databases: MS SQL Server
- Metabase version: v0.33.4
- Metabase hosting environment: Jar file on Windows server
- Metabase internal database: Postgres

**Severity**
Frustrating and stopping me from using the advanced query builder to it's full potential. If I can't change previous steps then what's the point?

### PR DESCRIPTION
**Pull request by tlrobinson**
Also adds tests/docs/flow types for steps.js

Resolves #11334

### CODE REVIEW

File: frontend/src/metabase-lib/lib/queries/StructuredQuery.js [hunk 1]
```diff
@@ -1350,6 +1362,15 @@ export default class StructuredQuery extends AtomicQuery {
     return db && db.hasFeature("nested-queries");
   }
 
+  cleanEmptyNesting(): StructuredQuery {
```
Comment (paulrosenzweig) [inline frontend/src/metabase-lib/lib/queries/StructuredQuery.js]: Looks like `cleanNesting` replaced this?

File: frontend/src/metabase-lib/lib/queries/StructuredQuery.js [hunk 2]
```diff
@@ -351,58 +351,70 @@ export default class StructuredQuery extends AtomicQuery {
...
-      .cleanFields();
+      .cleanFields()
+      .cleanEmpty();
+  }
 
-    // remove empty queries
-    const newSourceQuery = query.sourceQuery();
-    if (newSourceQuery && (query.isEmpty() || !query.hasAnyClauses())) {
-      return newSourceQuery;
+  cleanNesting(): StructuredQuery {
```
Comment (paulrosenzweig) [inline frontend/src/metabase-lib/lib/queries/StructuredQuery.js]: Pretty simple, but maybe worth a unit test.

File: frontend/src/metabase/query_builder/components/notebook/lib/steps.js [hunk 3]
```diff
@@ -1,8 +1,86 @@
...
+  query: StructuredQuery,
+  // a query to preview query at this step
+  previewQuery: ?StructuredQuery,
+  // remove this step
+  revert: ?(query: StructuredQuery) => StructuredQuery,
+  // remove invalid clauses set by this step
+  clean: (query: StructuredQuery) => StructuredQuery,
+  // update the query at this step and clean subsequent queries
+  update: (datasetQuery: StructuredDatasetQuery) => StructuredQuery,
+  // any valid actions that can be applied after this step
```
Comment (paulrosenzweig) [inline frontend/src/metabase/query_builder/components/notebook/lib/steps.js]: These are great comments! They're concise, but they add a lot over the unexplained method names.

**Review by paulrosenzweig — APPROVED** [review]
I don't understand which part fixed the bug, but all the changes look good to me.

File: frontend/src/metabase-lib/lib/queries/StructuredQuery.js (hunk 1 above)
Comment (tlrobinson) [reply frontend/src/metabase-lib/lib/queries/StructuredQuery.js]: Yup good catch

File: frontend/src/metabase-lib/lib/queries/StructuredQuery.js (hunk 2 above)
Comment (tlrobinson) [reply frontend/src/metabase-lib/lib/queries/StructuredQuery.js]: 👍 Thanks for keeping me honest :)

**Review by mazameli — APPROVED** [review]
Seems to fix the repro case spelled out in https://github.com/metabase/metabase/issues/11334

**Comment by mazameli** [comment]
Shouldn't the target for this PR be the release branch since it's a bug fix?

**Comment by tlrobinson** [comment]
> Shouldn't the target for this PR be the release branch since it's a bug fix?

I'm going to go ahead and merge this and cherry-pick into another branch off release-0.34.x

**Pull request merged**